README.md
# MMM-Oura

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/). It displays sleep data from your Oura sleep tracker.

## Installation

1. Go to MagicMirror's `modules`.
2. Run `git clone https://github.com/erchenger/MMM-Oura`. 
3. Go into the newly created directory: `cd MMM-Oura`.
4. Run `npm install` to install the node dependencies.

To use this module, add the following configuration block to the modules array in the `config/config.js` file:

```js
var config = {
  modules: [
    {
      module: "MMM-Oura",
      position: 'top_right', // Or any other region
      config: {
        // See below for configurable options
      }
    }
  ]
};
```

## Screenshot

![](MMM-Oura-screenshot.png)
## Configuration options

| Option    | Description                                                                                                     |
| --------- | --------------------------------------------------------------------------------------------------------------- |
| `apiKey` | _Required_ <b>Your [Oura personal access token](https://cloud.ouraring.com/docs/authentication#personal-access-tokens)</b>                                                                                   |
| `interval` | _Optional_ Interval in seconds how often the data should be fetched <br><br>**Type:** `int`(seconds) <br>Default 3600 seconds (1 hour) |


CHANGELOG.md
# MMM-Oura Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - Unreleased

First public release


MMM-Oura.js
/* global Module */

/* Magic Mirror
 * Module: MMM-Oura
 *
 * By Elliott Chenger
 * MIT Licensed.
 */

Module.register("MMM-Oura", {
  defaults: {
    interval: 30,
    apiKey: ""
  },

  requiresVersion: "2.1.0", // Required version of MagicMirror
  socketNotificationReceived: function (notification, payload) {
    if (notification === "SLEEP") {
      this.sleep = payload;
      this.updateDom();
    }
  },

  start: function () {
    const self = this;
    self.sendSocketNotification("CONFIG", self.config);
    const seconds = this.config.interval * 1000;
    window.setInterval(function () {
      self.sendSocketNotification("CONFIG", self.config);
      self.sendSocketNotification("GET_SLEEP");
    }, seconds);
  },
  getDom: function () {
    var self = this;
    //Scores Row
    var scoresRow = document.createElement("tr");
    scoresRow.className = "oura-scores";

    //Readiness Score
    var readinessScoreContainer = document.createElement("th");
    var readinessScore = document.createTextNode(
      this.sleep.readinessScore || "?"
    );
    readinessScoreContainer.appendChild(readinessScore);
    scoresRow.appendChild(readinessScoreContainer);

    // Sleep Score
    var sleepScoreContainer = document.createElement("th");
    var sleepScore = document.createTextNode(this.sleep.sleepScore || "?");
    sleepScoreContainer.appendChild(sleepScore);
    scoresRow.appendChild(sleepScoreContainer);

    // Sleep Duration
    var sleepDurationContainer = document.createElement("th");
    var sleepDuration = document.createTextNode(
      this.sleep.sleepDuration || "?"
    );
    sleepDurationContainer.appendChild(sleepDuration);
    scoresRow.appendChild(sleepDurationContainer);

    //Labels Row
    var labelsRow = document.createElement("tr");
    labelsRow.className = "oura-lables";

    //Readiness Label
    var readinessLabelContainer = document.createElement("th");
    var readinessLabel = document.createTextNode("Readiness");
    readinessLabelContainer.appendChild(readinessLabel);
    labelsRow.appendChild(readinessLabelContainer);

    //Sleep Label
    var sleepLabelContainer = document.createElement("th");
    var sleepLabel = document.createTextNode("Sleep");
    sleepLabelContainer.appendChild(sleepLabel);
    labelsRow.appendChild(sleepLabelContainer);

    var sleepDurationLabelContainer = document.createElement("th");
    var sleepDurationLabel = document.createTextNode("Duration");
    sleepDurationLabelContainer.appendChild(sleepDurationLabel);
    labelsRow.appendChild(sleepDurationLabelContainer);

    var table = document.createElement("table");
    table.className = "oura-table";
    table.appendChild(scoresRow);
    table.appendChild(labelsRow);

    var wrapper = document.createElement("div");
    wrapper.appendChild(table);
    return wrapper;
  },

  getScripts: function () {
    return [];
  },

  getStyles: function () {
    return ["MMM-Oura.css"];
  },

  // Load translations files
  getTranslations: function () {
    //FIXME: This can be load a one file javascript definition
    return {
      en: "translations/en.json",
      es: "translations/es.json"
    };
  }
});


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-Oura
 *
 * By Elliott Chenger
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
const oura = require("oura");
const dateFormat = require("dateformat");

module.exports = NodeHelper.create({
  start: () => {
    this.config = {};
  },
  // Override socketNotificationReceived method.

  /* socketNotificationReceived(notification, payload)
   * This method is called when a socket notification arrives.
   *
   * argument notification string - The identifier of the noitication.
   * argument payload mixed - The payload of the notification.
   */
  socketNotificationReceived: function (notification, payload) {
    const self = this;
    if (notification === "CONFIG") {
      self.config = payload;
      self.sendSocketNotification("GET_SLEEP");
    }
    if (notification === "GET_SLEEP") {
      const today = new Date();
      const yesterday = new Date(today);
      yesterday.setDate(yesterday.getDate() - 1);
      const todayString = dateFormat(today, "yyyy-mm-dd");
      const yesterdayString = dateFormat(yesterday, "yyyy-mm-dd");
      const ouraClient = new oura.Client(self.config.apiKey);
      const sleepResonse = ouraClient.sleep(yesterdayString, todayString);
      const readinessResponse = ouraClient.readiness(
        yesterdayString,
        todayString
      );
      Promise.all([sleepResonse, readinessResponse])
        .then(([sleep, readiness]) => {
          const sleepDuration = sleep.sleep[0].total;
          const sleepHrDuration = Math.floor(sleepDuration / 3600);
          const sleepMinsDuration = (sleepDuration % 3600) / 60;
          const hrLabel = sleepHrDuration > 1 ? "hrs" : "hr";
          const minLabel = sleepMinsDuration > 1 ? "mins" : "min";
          const data = {
            sleepDuration: `${sleepHrDuration}${hrLabel} ${sleepMinsDuration}${minLabel}`,
            sleepScore: sleep.sleep[0].score || "?",
            readinessScore: readiness.readiness[0].score || "?"
          };
          self.sendSocketNotification("SLEEP", data);
        })
        .catch(function (error) {
          console.log(error);
        });
    }
  }
});


MagicMirrorModule-Oura.js
/* global Module */

/* Magic Mirror
 * Module: MMM-Oura
 *
 * By Elliott Chenger
 * MIT Licensed.
 */

Module.register("MMM-Oura", {
  defaults: {
    interval: 3600,
    apiKey: ""
  },

  requiresVersion: "2.1.0", // Required version of MagicMirror
  socketNotificationReceived: function (notification, payload) {
    if (notification === "SLEEP") {
      this.sleep = payload;
      this.updateDom();
    }
  },

  start: function () {
    const self = this;
    self.sendSocketNotification("CONFIG", self.config);
    const seconds = this.config.interval * 1000;
    window.setInterval(function () {
      self.sendSocketNotification("CONFIG", self.config);
      self.sendSocketNotification("GET_SLEEP");
    }, seconds);
  },
  getDom: function () {
    var self = this;
    //Scores Row
    var scoresRow = document.createElement("tr");
    scoresRow.className = "oura-scores";

    //Readiness Score
    var readinessScoreContainer = document.createElement("th");
    var readinessScore = document.createTextNode(
      this.sleep.readinessScore || "?"
    );
    readinessScoreContainer.appendChild(readinessScore);
    scoresRow.appendChild(readinessScoreContainer);

    // Sleep Score
    var sleepScoreContainer = document.createElement("th");
    var sleepScore = document.createTextNode(this.sleep.sleepScore || "?");
    sleepScoreContainer.appendChild(sleepScore);
    scoresRow.appendChild(sleepScoreContainer);

    // Sleep Duration
    var sleepDurationContainer = document.createElement("th");
    var sleepDuration = document.createTextNode(
      this.sleep.sleepDuration || "?"
    );
    sleepDurationContainer.appendChild(sleepDuration);
    scoresRow.appendChild(sleepDurationContainer);

    //Labels Row
    var labelsRow = document.createElement("tr");
    labelsRow.className = "oura-lables";

    //Readiness Label
    var readinessLabelContainer = document.createElement("th");
    var readinessLabel = document.createTextNode("Readiness");
    readinessLabelContainer.appendChild(readinessLabel);
    labelsRow.appendChild(readinessLabelContainer);

    //Sleep Label
    var sleepLabelContainer = document.createElement("th");
    var sleepLabel = document.createTextNode("Sleep");
    sleepLabelContainer.appendChild(sleepLabel);
    labelsRow.appendChild(sleepLabelContainer);

    var sleepDurationLabelContainer = document.createElement("th");
    var sleepDurationLabel = document.createTextNode("Duration");
    sleepDurationLabelContainer.appendChild(sleepDurationLabel);
    labelsRow.appendChild(sleepDurationLabelContainer);

    var table = document.createElement("table");
    table.className = "oura-table";
    table.appendChild(scoresRow);
    table.appendChild(labelsRow);

    var wrapper = document.createElement("div");
    wrapper.appendChild(table);
    return wrapper;
  },

  getScripts: function () {
    return [];
  },

  getStyles: function () {
    return ["MMM-Oura.css"];
  },

  // Load translations files
  getTranslations: function () {
    //FIXME: This can be load a one file javascript definition
    return {
      en: "translations/en.json",
      es: "translations/es.json"
    };
  }
});
